src/useLocalStorage.ts

```diff
@@ -0,0 +1,82 @@
+import { LocalStorage } from "@raycast/api";
+import { useCachedPromise } from "./useCachedPromise";
+import { showFailureToast } from "./showFailureToast";
+import { replacer, reviver } from "./helpers";
+
+export type UseLocalStorageReturnValue<T> = {
+  value: T;
+  setValue: (value: T) => Promise<void>;
+  removeValue: () => Promise<void>;
+  isLoading: boolean;
+};
+
+export type UseLocalStorageReturnValueWithUndefined<T> = {
+  value?: T;
+  setValue: (value: T) => Promise<void>;
+  removeValue: () => Promise<void>;
+  isLoading: boolean;
+};
+
+/**
+ * A hook to manage a value in the local storage.
+ *
+ * @remark The value is stored as a JSON string in the local storage.
+ *
+ * @param key - The key to use for the value in the local storage.
+ * @param initialValue - The initial value to use if the key doesn't exist in the local storage.
+ * @returns An object with the following properties:
+ * - `value`: The value from the local storage or the initial value if the key doesn't exist.
+ * - `setValue`: A function to update the value in the local storage.
+ * - `removeValue`: A function to remove the value from the local storage.
+ * - `isLoading`: A boolean indicating if the value is loading.
+ *
+ * @example
+ * ```
+ * const { value, setValue } = useLocalStorage<string>("my-key");
+ * const { value, setValue } = useLocalStorage<string>("my-key", "default value");
+ * ```
+ */
+export function useLocalStorage<T>(key: string, initialValue: T): UseLocalStorageReturnValue<T>;
+export function useLocalStorage<T>(key: string): UseLocalStorageReturnValueWithUndefined<T>;
+export function useLocalStorage<T>(key: string, initialValue?: T) {
+  const {
+    data: value,
+    isLoading,
+    mutate,
+  } = useCachedPromise(
+    async (storageKey: string) => {
+      const item = await LocalStorage.getItem<string>(storageKey);
+
+      if (item) {
+        return JSON.parse(item, reviver);
+      }
+    },
+    [key],
+  );
+
+  async function setValue(value: T) {
+    try {
+      await mutate(LocalStorage.setItem(key, JSON.stringify(value, replacer)), {
+        optimisticUpdate() {
+          return value;
+        },
+      });
+    } catch (error) {
+      await showFailureToast(error, { title: "Failed to set value in local storage" });
+    }
+  }
+
+  async function removeValue() {
+    try {
+      await mutate(LocalStorage.removeItem(key), {
+        optimisticUpdate() {
+          return undefined;
+        },
+      });
+    } catch (error) {
+      await showFailureToast(error, { title: "Failed to remove value from local storage" });
+    }
+  }
+
+  return { value: value ?? initialValue, setValue, removeValue, isLoading };
```

Comment:
Any reason to use the `initialValue` here instead of passing it to the `useCachedState`?

Thinking about it more, I think that when the cache is wiped (which can happen at any time) the initialValue will flicker even if there's something on the localStorage.

I'm wondering if we actually want the `useCachedPromise` and the `initialValue` 🤔

Reply:
Did you mean `useCachedPromise` instead of `useCachedState`? The reason is to avoid having `undefined` returned even if the user passed an initial value. What would you suggest instead?

Reply:
yes sorry.

> The reason is to avoid having undefined returned even if the user passed an initial value.

I don't understand, `useCachedPromise` should do that already.

But yeah I don't think caching the value and returning the initial value if there is no cache is what we want. I understand that it would allow you to get the previous value (maybe, sometimes) without having to wait but it will lead to flickers

Reply:
> I don't understand, useCachedPromise should do that already.

I don't know if that's intended or not, but even if I passed the `initialData` option to `useCachedPromise`, the value can still be `undefined` for a brief moment. I'd expect it to never be `undefined` if there's an initial value 🤔

So should we go with `usePromise` then?

Reply:
> I don't know if that's intended or not, but even if I passed the initialData option to useCachedPromise, the value can still be undefined for a brief moment

Hum that's a bug then. Do you have a repro?

> So should we go with usePromise then?

I'd say so yeah

Reply:
But, won't we get flickers as well with `usePromise`? If an initial value is passed, the data will always load and we can get flickers as well if the value in the local storage is different than the initial one. Here's an example where the initial value is a list of three to-dos:

https://github.com/raycast/utils/assets/16003285/97e9b87d-5cc5-48c9-b041-94336e47ccc4

Unless we should get rid of `initialValue` as well?

Reply:
> I'm wondering if we actually want the useCachedPromise and the initialValue

That's what I said yeah, I don't think we need either